File: __tests__/lib/week.test.ts
// Tests des aides « semaine » (lib/week.ts) : début de semaine (lundi), clé de
// jour locale et agrégation des durées en barres. Pur. Les dates sont
// construites en heure locale pour rester indépendant du fuseau de la CI.
import { dailyDurationBars, localDayKey, startOfWeekMs } from '@/lib/week';

describe('startOfWeekMs', () => {
  it('renvoie le lundi minuit (local) de la semaine', () => {
    // mercredi 11 juin 2025, 14:30 local
    const now = new Date(2025, 5, 11, 14, 30).getTime();
    const monday = new Date(startOfWeekMs(now));
    expect(monday.getFullYear()).toBe(2025);
    expect(monday.getMonth()).toBe(5);
    expect(monday.getDate()).toBe(9); // lundi 9 juin
    expect(monday.getDay()).toBe(1); // lundi
    expect(monday.getHours()).toBe(0);
    expect(monday.getMinutes()).toBe(0);
  });

  it('un lundi se renvoie lui-même (minuit)', () => {
    const monday = new Date(2025, 5, 9, 8, 0).getTime();
    const start = new Date(startOfWeekMs(monday));
    expect(start.getDate()).toBe(9);
    expect(start.getHours()).toBe(0);
  });

  it('un dimanche renvoie le lundi précédent', () => {
    const sunday = new Date(2025, 5, 15, 23, 0).getTime(); // dimanche
    const start = new Date(startOfWeekMs(sunday));
    expect(start.getDate()).toBe(9); // lundi 9
    expect(start.getDay()).toBe(1);
  });
});

describe('localDayKey', () => {
  it('formate en YYYY-MM-DD avec zéro initial', () => {
    expect(localDayKey(new Date(2025, 0, 5))).toBe('2025-01-05');
    expect(localDayKey(new Date(2025, 11, 31))).toBe('2025-12-31');
  });
});

describe('dailyDurationBars', () => {
  const now = new Date(2025, 5, 11, 10, 0).getTime(); // mercredi 11 juin

  it('produit `days` barres, aujourd’hui en dernier', () => {
    const bars = dailyDurationBars([], 7, now);
    expect(bars).toHaveLength(7);
    expect(bars.every((b) => b.value === 0)).toBe(true);
    // aujourd'hui = mercredi -> dernière barre labellisée 'M'
    expect(bars[6].label).toBe('M');
  });

  it('apparie les durées par clé de jour locale', () => {
    const daily = [
      { day: localDayKey(new Date(2025, 5, 11)), durationSec: 1800 }, // aujourd'hui
      { day: localDayKey(new Date(2025, 5, 9)), durationSec: 3600 }, // lundi
    ];
    const bars = dailyDurationBars(daily, 7, now);
    expect(bars[6].value).toBe(1800); // mercredi (dernier)
    expect(bars[4].value).toBe(3600); // lundi (5 jours plus tôt -> index 4 sur 7)
    expect(bars[0].label).toBe('J'); // 6 jours avant mercredi = jeudi précédent
  });

  it('ignore les jours hors fenêtre', () => {
    const daily = [{ day: localDayKey(new Date(2025, 5, 1)), durationSec: 9999 }];
    const bars = dailyDurationBars(daily, 7, now);
    expect(bars.every((b) => b.value === 0)).toBe(true);
  });
});

File: app/src/test/java/ovh/battistella/elan/domain/WeekTest.kt
// Tests des aides « semaine » : début de semaine (lundi), clé de jour locale et
// agrégation des durées en barres. Dates construites en heure locale pour
// rester indépendant du fuseau de la CI.
package ovh.battistella.elan.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId

class WeekTest {
    private fun local(year: Int, month: Int, day: Int, hour: Int = 0, minute: Int = 0): Long =
        LocalDateTime.of(year, month, day, hour, minute).atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()

    private fun zoned(ms: Long) = Instant.ofEpochMilli(ms).atZone(ZoneId.systemDefault())

    @Test
    fun `startOfWeekMs renvoie le lundi minuit (local) de la semaine`() {
        // mercredi 11 juin 2025, 14:30 local
        val monday = zoned(startOfWeekMs(local(2025, 6, 11, 14, 30)))
        assertEquals(2025, monday.year)
        assertEquals(6, monday.monthValue)
        assertEquals(9, monday.dayOfMonth) // lundi 9 juin
        assertEquals(DayOfWeek.MONDAY, monday.dayOfWeek)
        assertEquals(0, monday.hour)
        assertEquals(0, monday.minute)
    }

    @Test
    fun `startOfWeekMs un lundi se renvoie lui-même (minuit)`() {
        val start = zoned(startOfWeekMs(local(2025, 6, 9, 8, 0)))
        assertEquals(9, start.dayOfMonth)
        assertEquals(0, start.hour)
    }

    @Test
    fun `startOfWeekMs un dimanche renvoie le lundi précédent`() {
        val start = zoned(startOfWeekMs(local(2025, 6, 15, 23, 0))) // dimanche
        assertEquals(9, start.dayOfMonth)
        assertEquals(DayOfWeek.MONDAY, start.dayOfWeek)
    }

    @Test
    fun `localDayKey formate en YYYY-MM-DD avec zéro initial`() {
        assertEquals("2025-01-05", localDayKey(LocalDate.of(2025, 1, 5)))
        assertEquals("2025-12-31", localDayKey(LocalDate.of(2025, 12, 31)))
        assertEquals("2025-01-05", localDayKey(local(2025, 1, 5, 23, 59)))
    }

    private val now = local(2025, 6, 11, 10, 0) // mercredi 11 juin

    @Test
    fun `dailyDurationBars produit days barres, aujourd'hui en dernier`() {
        val bars = dailyDurationBars(emptyList(), 7, now)
        assertEquals(7, bars.size)
        assertTrue(bars.all { it.value == 0 })
        // aujourd'hui = mercredi -> dernière barre labellisée 'M'
        assertEquals("M", bars[6].label)
    }

    @Test
    fun `dailyDurationBars apparie les durées par clé de jour locale`() {
        val daily = listOf(
            DailyDuration(localDayKey(LocalDate.of(2025, 6, 11)), 1800), // aujourd'hui
            DailyDuration(localDayKey(LocalDate.of(2025, 6, 9)), 3600), // lundi
        )
        val bars = dailyDurationBars(daily, 7, now)
        assertEquals(1800, bars[6].value) // mercredi (dernier)
        assertEquals(3600, bars[4].value) // lundi (5 jours plus tôt -> index 4 sur 7)
        assertEquals("J", bars[0].label) // 6 jours avant mercredi = jeudi précédent
    }

    @Test
    fun `dailyDurationBars ignore les jours hors fenêtre`() {
        val daily = listOf(DailyDuration(localDayKey(LocalDate.of(2025, 6, 1)), 9999))
        assertTrue(dailyDurationBars(daily, 7, now).all { it.value == 0 })
    }
}
